File: src/main/java/com/github/syuchan1005/imagecomment/ImageLineMarkerProvider.java
package com.github.syuchan1005.imagecomment;

import com.intellij.codeHighlighting.Pass;
import com.intellij.codeInsight.daemon.LineMarkerInfo;
import com.intellij.codeInsight.daemon.LineMarkerProvider;
import com.intellij.openapi.editor.markup.GutterIconRenderer;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import java.net.URL;
import java.util.Collection;
import java.util.List;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ImageLineMarkerProvider implements LineMarkerProvider {
	@Nullable
	@Override
	public LineMarkerInfo getLineMarkerInfo(@NotNull PsiElement element) {
		return null;
	}

	@Override
	public void collectSlowLineMarkers(@NotNull List<PsiElement> elements, @NotNull Collection<LineMarkerInfo> result) {
		elements.forEach(element -> {
			String typeName = element.getNode().getElementType().toString();
			if (!typeName.toUpperCase().endsWith("COMMENT")) return;

			CommentUtil.getMatchData(element.getText()).forEach(match -> {
				URL url = CommentUtil.getAbsoluteURL(element, match.getPath());
				assert url != null;

				int startOffset = element.getTextRange().getStartOffset() + match.startOffset;
				CommentUtil.ImageData imageData = CommentUtil.loadImage(url);
				if (imageData == null) return;
				LineMarkerInfo lineMarkerInfo = new LineMarkerInfo<>(element,
						new TextRange(startOffset, startOffset + match.text.length()),
						imageData.toIcon(), Pass.LINE_MARKERS,
						null, null, GutterIconRenderer.Alignment.CENTER);

				result.add(lineMarkerInfo);
			});
		});
	}


}

File: src/main/kotlin/com/github/syuchan1005/imagecomment/ImageLineMarkerProvider.kt
package com.github.syuchan1005.imagecomment

import com.intellij.codeInsight.daemon.RelatedItemLineMarkerInfo
import com.intellij.codeInsight.daemon.RelatedItemLineMarkerProvider
import com.intellij.codeInsight.navigation.NavigationGutterIconBuilder
import com.intellij.psi.PsiElement
import com.intellij.util.IconUtil
import com.intellij.util.ui.ImageUtil
import javax.imageio.ImageIO
import kotlin.math.max

class ImageLineMarkerProvider : RelatedItemLineMarkerProvider() {
    override fun collectNavigationMarkers(
        element: PsiElement,
        result: MutableCollection<in RelatedItemLineMarkerInfo<*>>
    ) {
        if (!element.node.elementType.toString().uppercase().endsWith("COMMENT")) {
            return
        }

        CommentUtil.getMatchData(element.text).forEach { matchData ->
            val bytes = CommentUtil.getImageByteArray(element, matchData.path) ?: return@forEach
            val bufferedImage = bytes.inputStream().use(ImageIO::read) ?: return@forEach
            val scaledImage = ImageUtil.scaleImage(
                bufferedImage,
                ICON_SIZE / max(bufferedImage.width, bufferedImage.height)
            )

            val builder = NavigationGutterIconBuilder.create(IconUtil.createImageIcon(scaledImage))
                .setTargets(element)
                .setTooltipText(matchData.alt)
            result.add(
                builder.createLineMarkerInfo(element)
            )
        }
    }

    private companion object {
        const val ICON_SIZE: Double = 16.0
    }
}
